src/xenia/vfs/devices/stfs_container_device.h

```diff
@@ -24,144 +27,390 @@ namespace vfs {
-  char16_t display_descs[0x900 / 2];
-  char16_t publisher_name[0x80 / 2];
-  char16_t title_name[0x80 / 2];
-  uint8_t transfer_flags;
-  uint32_t thumbnail_image_size;
-  uint32_t title_thumbnail_image_size;
-  uint8_t thumbnail_image[0x4000];
-  uint8_t title_thumbnail_image[0x4000];
-
-  // Metadata v2 Fields
-  uint8_t series_id[0x10];
-  uint8_t season_id[0x10];
-  int16_t season_number;
-  int16_t episode_number;
-  char16_t additonal_display_names[0x300 / 2];
-  char16_t additional_display_descriptions[0x300 / 2];
+  union {
+    be<uint16_t> display_name_raw[9][128];
+    char16_t display_name_chars[9][128];
+  };
+  union {
+    be<uint16_t> description_raw[9][128];
+    char16_t description_chars[9][128];
+  };
+  union {
+    be<uint16_t> publisher_raw[64];
+    char16_t publisher_chars[64];
+  };
+  union {
+    be<uint16_t> title_name_raw[64];
+    char16_t title_name_chars[64];
+  };
+  union {
+    XContentAttributes bits;
+    uint8_t as_byte;
+  } flags;
+  be<uint32_t> thumbnail_size;
+  be<uint32_t> title_thumbnail_size;
+  uint8_t thumbnail[0x3D00];
+  union {
+    be<uint16_t> display_name_ex_raw[3][128];
+    char16_t display_name_ex_chars[3][128];
+  };
+  uint8_t title_thumbnail[0x3D00];
+  union {
+    be<uint16_t> description_ex_raw[3][128];
+    char16_t description_ex_chars[3][128];
+  };
+
+  std::u16string display_name(XLanguage language) const {
+    if (language >= XLanguage::kMaxLanguages) {
+      assert_always();
+      // no room for this lang, read from english slot..
+      language = XLanguage::kEnglish;
+    }
+
+    uint32_t lang_id = uint32_t(language) - 1;
+
+    const be<uint16_t>* str = 0;
+    if (lang_id >= 0 && lang_id < 9) {
+      str = display_name_raw[lang_id];
+    } else if (lang_id >= 9 && lang_id < 12 && metadata_version >= 2) {
+      str = display_name_ex_raw[lang_id - 9];
+    }
+
+    if (!str) {
+      // Invalid language ID?
+      assert_always();
+      return u"";
+    }
+
+    return load_and_swap<std::u16string>(str);
+  }
+
+  std::u16string description(XLanguage language) const {
+    if (language >= XLanguage::kMaxLanguages) {
+      assert_always();
+      // no room for this lang, read from english slot..
+      language = XLanguage::kEnglish;
+    }
+
+    uint32_t lang_id = uint32_t(language) - 1;
+
+    const be<uint16_t>* str = 0;
+    if (lang_id >= 0 && lang_id < 9) {
+      str = description_raw[lang_id];
+    } else if (lang_id >= 9 && lang_id < 12 && metadata_version >= 2) {
+      str = description_ex_raw[lang_id - 9];
+    }
+
+    if (!str) {
+      // Invalid language ID?
+      assert_always();
+      return u"";
+    }
+
+    return load_and_swap<std::u16string>(str);
+  }
+
+  std::u16string publisher() const {
+    return load_and_swap<std::u16string>(publisher_raw);
+  }
+
+  std::u16string title_name() const {
+    return load_and_swap<std::u16string>(title_name_raw);
+  }
+
+  bool display_name(XLanguage language, const std::u16string_view value) {
```

Comment:
If I'm understanding the context here, if this is updating values shouldn't they be ie `set_display_name()` rather than just `display_name()`?

Reply:
Sure, just named them like that to make them sorta feel like actual fields, but I can rename them, would probably help make code that uses them a little clearer too.